frontend/discourse/app/static/prosemirror/components/image-node-view.gjs

```diff
@@ -276,11 +307,261 @@ export default class ImageNodeView extends Component {
+      null,
+      state.schema.nodes.paragraph.create(null, imageNode)
+    );
+
+    let gridStartPos;
+    if (isOnlyChildInParagraph) {
+      tr.replaceWith(pos - 1, pos + 1, gridNode);
+      gridStartPos = pos - 1;
+    } else {
+      tr.replaceWith(pos, pos + 1, gridNode);
+      gridStartPos = pos;
+    }
+
+    // Select the image inside the new grid: grid + paragraph + image
+    // Structure: grid(1) > paragraph(1) > image
+    const imagePos = gridStartPos + 2; // Skip grid node boundary and paragraph node boundary
+    tr.setSelection(NodeSelection.create(tr.doc, imagePos)).scrollIntoView();
+
+    view.dispatch(tr);
+  }
+
+  /**
+   * Check if the image at the resolved position is the only child in its paragraph.
+   * Uses position-based API for reliable detection instead of direct node comparison.
+   *
+   * @param {import("prosemirror-model").ResolvedPos} $pos - The resolved position of the image
+   * @returns {boolean} True if the image is the only child in a paragraph
+   */
+  #isImageOnlyChildInParagraph($pos) {
+    const parent = $pos.parent;
+
+    // Must be in a paragraph
+    if (parent.type.name !== "paragraph") {
+      return false;
+    }
+
+    // Paragraph must have exactly one child
+    if (parent.childCount !== 1) {
+      return false;
+    }
+
+    // The single child must be an image at the current position
+    const index = $pos.index();
+    const childAtIndex = parent.child(index);
+
+    return childAtIndex.type.name === "image";
+  }
+
+  #findNearbyGrid(pos) {
+    const { state } = this.args.view;
+    const $pos = state.doc.resolve(pos);
+
+    // Find the block containing the image (paragraph)
+    let blockDepth = $pos.depth;
+    while (blockDepth > 0 && $pos.node(blockDepth).type.name !== "paragraph") {
+      blockDepth--;
+    }
+
+    if (blockDepth === 0) {
+      return null;
+    }
+
+    const blockPos = $pos.start(blockDepth);
+    const blockNode = $pos.node(blockDepth);
+    const blockParent = $pos.node(blockDepth - 1);
+    const blockIndex = $pos.index(blockDepth - 1);
+
+    // Check left sibling
+    if (blockIndex > 0) {
+      const leftNode = blockParent.child(blockIndex - 1);
+      if (leftNode.type.name === "grid") {
+        return {
+          node: leftNode,
+          pos: blockPos - leftNode.nodeSize,
+          direction: -1,
+        };
+      }
+    }
+
+    // Check right sibling
+    if (blockIndex < blockParent.childCount - 1) {
+      const rightNode = blockParent.child(blockIndex + 1);
+      if (rightNode.type.name === "grid") {
+        return {
+          node: rightNode,
+          pos: blockPos + blockNode.nodeSize,
+          direction: 1,
+        };
+      }
+    }
+
+    return null;
+  }
+
+  @action
+  moveOutsideGrid() {
+    if (!this.isInGrid) {
+      return;
+    }
+
+    const pos = this.args.getPos();
+    const view = this.args.view;
+    const { state } = view;
+    const { tr } = state;
```

Comment:
Up to you, but this seems repeated a lot. Maybe we can have a function like:

```
get viewState() {
    const pos = this.args.getPos();
    const view = this.args.view;
    const { state } = view;
    const { tr } = state;

    return {
      pos,
      view,
      state,
      tr
    };
}
```

Then the caller can do `const { pos, view, state, tr } = this.viewState`  ?

Reply:
I saw only 2 places needing all of them, and this seems an unnecessary indirection IMHO, I think it's clearer with the direct uses. We might add getters for `pos`, `view`, `state` and use `this.x` directly instead of te local convenience variables, but it can be refactored on another PR